lib/semantics/resolve-names.cc

```diff
@@ -1782,19 +2050,30 @@ void ModuleVisitor::SetAccess(const parser::Name &name, Attr attr) {
   }
 }
 
+static bool HasExplicitType(const Symbol &symbol) {
+  if (symbol.has<UnknownDetails>()) {
+    return false;
+  } else if (const auto *details = symbol.detailsIf<EntityDetails>()) {
+    return details->type().has_value();
+  } else if (const auto *details = symbol.detailsIf<ObjectEntityDetails>()) {
+    return details->type().has_value();
+  } else if (const auto *details = symbol.detailsIf<ProcEntityDetails>()) {
+    return details->interface().symbol() != nullptr ||
+        details->interface().type() != nullptr;
+  } else {
+    return true;  // doesn't need explicit type
```

Comment:
Consider reversing the sense of this predicate, maybe to StillNeedsExplicitType(), to avoid the confusion of returning true for things that don't actually have types, just to avoid future confusion when somebody writes some code that crashes if HasExplicitType is true and yet later there is no type to be found in the object.

Reply:
Yes, that's much clearer.

Reply:
LLVM coding convention: don't use if () { return; }  else if () { return; } else return; but rather if () { return; } if () return; ... return;

Reply:
Thanks, Eric.  We sometimes violate that convention when every branch of a conditional construct, including its final `else` block, has a `return` statement, since the whole construct can be viewed as a big expression determining the return value, rather than one that uses `return` for control flow purposes like `break`.